### PR TITLE
Jb expr sortby

### PR DESCRIPTION
**Pull request by jbloom22**
Once this is in, I'll add a straight-up sort, and perhaps also options for descending rather than ascending order.

I"m still thinking about whether null should be at the beginning or the end, but we'll start with it at the beginning.

### CODE REVIEW

File: src/main/scala/org/broadinstitute/hail/Utils.scala [hunk 1]
```diff
@@ -1290,4 +1290,18 @@ object Utils extends Logging {
...
+  def extendOrderingToNull[T](ord: Ordering[T]): Ordering[Any] = {
+    object newOrd extends Ordering[Any] {
+      def compare(a: Any, b: Any) =
+        if (a != null)
+          if (b != null)
+            ord.compare(a.asInstanceOf[T], b.asInstanceOf[T])
+          else
+            1
+        else
+          -1
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/Utils.scala]: This needs to return 0 if they are both null.

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/Utils.scala]: You can just write

```
  def extendOrderingToNull[T](ord: Ordering[T]): Ordering[Any] = {
    new Ordering[Any] {
      def compare(a: Any, b: Any) = ...
    }
  }
```

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 2]
```diff
@@ -839,6 +839,18 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
           parseError(s"method `$method' expects a lambda function [param => Boolean], got [param => ${body.`type`}]")
         `type` = TBoolean
 
+      case (arr: TArray, "sortBy", rhs) =>
+        lhs.typecheck(ec)
+        val (param, body) = rhs match {
+          case Array(Lambda(_, p, b)) => (p, b)
+          case _ => parseError(s"method `$method' expects a lambda function [param => T] with T of string or numeric type, " +
+            s"e.g. `x => x.length'") // FIXME: better examples
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: Without context (what is x?) this is more confusing than not.  I don't think error messages are the place for examples.

**Comment by cseed** [comment]
Back to you.

**Comment by jbloom22** [comment]
back to you

**Comment by jbloom22** [comment]
squashed to 1 commit.  back to you

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 3]
```diff
@@ -839,6 +839,17 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
 
+      case (arr: TArray, "sortBy", rhs) =>
+        lhs.typecheck(ec)
+        val (param, body) = rhs match {
+          case Array(Lambda(_, p, b)) => (p, b)
+          case _ => parseError(s"method `$method' expects a lambda function [param => T] with T of string or numeric type")
+        }
+        body.typecheck(ec.copy(st = ec.st + ((param, (-1, arr.elementType)))))
+        if (!(body.`type`.isInstanceOf[TNumeric] || body.`type` == TString))
+          parseError(s"method `$method' expects a lambda function [param => T] with T of string or numeric type, got [param => ${body.`type`}]")
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: You didn't address my comment about the wording of this message.  Any reason why?

File: src/main/scala/org/broadinstitute/hail/Utils.scala [hunk 4]
```diff
@@ -1290,4 +1290,16 @@ object Utils extends Logging {
   }
 
   def uriPath(uri: String): String = new URI(uri).getPath
+
+  def extendOrderingToNull[T](ord: Ordering[T]): Ordering[Any] = {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/Utils.scala]: I'd make ord implicit.  Also, you don't need the outermost curly braces.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 5]
```diff
@@ -1067,6 +1077,25 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
+    case (returnType, "sortBy", Array(Lambda(_, param, body))) =>
+      val localIdx = ec.a.length
+      val localA = ec.a
+      val ord = (body.`type`: @unchecked) match {
+        case TDouble => extendOrderingToNull[Double](Ordering.Double)
+        case TFloat => extendOrderingToNull[Float](Ordering.Float)
+        case TLong => extendOrderingToNull[Long](Ordering.Long)
+        case TInt => extendOrderingToNull[Int](Ordering.Int)
+        case TString => extendOrderingToNull[String](Ordering.String)
+      }
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: If ordering is implicit, you don't need the `Ordering` arguments here.

**Comment by jbloom22** [comment]
made ordering implicit, back to you

**Comment by cseed** [comment]
jenkins

**Pull request merged**